[Core/UI] Add Model Debug Panel, Morph Conversion, and OpenCC Support

rintrint opened the pull request:
- Add Model Debug Panel for detecting issues with model/bone names, morphs, and textures
- Implement morph group to shape key conversion functionality
- Add optional dependency management system with OpenCC integration
- Allow users to decide which optional dependencies to install directly from the preferences panel

To add new optional dependencies, simply add an entry to the OPTIONAL_DEPENDENCIES dictionary: 
```
OPTIONAL_DEPENDENCIES = {
    "new_dep": {
        "display_name": "Display Name",
        "description": "What it does",
        "package_name": "pip_package_name",
        "module_name": "import_name",
        "icon": "BLENDER_ICON"
    }
}
```

Resolves #211, #168
Reference: https://github.com/BYVoid/OpenCC

Model Debug Panel Screenshot:
<img width="360" alt="Snipaste_2025-05-05_04-41-43" src="https://github.com/user-attachments/assets/ed2f5f0e-2d26-48a6-9047-6f1ebed7b04a" />

Merge Group Vertex Morphs Button Screenshot:
<img width="421" alt="Snipaste_2025-05-05_04-42-25" src="https://github.com/user-attachments/assets/7316f1c3-386c-4fe8-9efd-ef57e2594f52" />

Optional Dependencies Screenshot:
<img width="496" alt="Snipaste_2025-05-05_04-38-42" src="https://github.com/user-attachments/assets/96cccd7c-dbdd-4a75-9835-1d9684d29a14" />

nagadomi commented:
As you may know, bone morphs cannot be perfectly converted to vertex morphs.
Since shape keys move vertices in a straight line between the start and end points, the intermediate shapes may not match the expected result.
There may be cases where you may want to do that, but it is generally not a recommended process. 

It’s not enough to block the implementation, but it may require a warning or clarification.

rintrint commented:
> bone morphs cannot be perfectly converted to vertex morphs.
I didn't implement a conversion from bone morphs to vertex morphs.
This PR adds a button to combine multiple vertex morphs into a single vertex morph, which only addresses the issue described in #168.

nagadomi commented:
@rintrint Thanks for clarifying. Sorry for the hasty assumption.

UuuNyaa commented on mmd_tools/operators/model_validation.py:
```diff
@@ -0,0 +1,708 @@
+    is_opencc_available = False
+
+# Scene property to store validation results
+bpy.types.Scene.mmd_validation_results = StringProperty(
```
It would be better to implement register/unregister in static methods of the Operator class that depends on the property.

see: https://github.com/MMD-Blender/blender_mmd_tools/blob/main/mmd_tools/properties/rigid_body.py#L228

rintrint replied:
I'll use module-level register/unregister functions instead of static methods within operator classes since the property is shared across multiple operators. This approach works with auto_load.py. Not sure if this approach is suitable?

UuuNyaa replied:
I think it is a good decision👌

UuuNyaa commented on mmd_tools/operators/model_validation.py:
```diff
@@ -0,0 +1,708 @@
+)
+
+
+def log_to_console(prefix, message):
```
Have you considered using `logging`?

see: https://github.com/MMD-Blender/blender_mmd_tools/blob/main/mmd_tools/core/pmd/importer.py#L30

UuuNyaa commented:
@rintrint Thanks for your awesome contribution! 😃

I think we should remove our dependency on OpenCC for the following reasons:

1. The pip installation of unofficial Blender Python modules should be avoided
2. We want to avoid language-specific (Chinese in this case) features
    - how about improving [the global translation panel](https://github.com/MMD-Blender/blender_mmd_tools/blob/main/mmd_tools/operators/translations.py#L224) and preset translation dictionaries instead?
3. How about implementing it as a separate extension if it is necessary?

rintrint commented:
@UuuNyaa Thank you for your feedback, UuuNyaa! I'd like to clarify some points about the OpenCC implementation.

The primary purpose of adding OpenCC is not to support Chinese, but rather to address ShiftJIS encoding issues in VMD files. This helps fix Japanese character typos that commonly occur in morph names (e.g., '口横缩げ' -> '口横縮げ'). While creating a new format supporting UTF16LE/UTF8 would be ideal, that's a much larger undertaking.

Regarding your concerns:

1. About pip installation: I understand the concerns about unofficial modules. That's why I implemented this as an optional dependency that users can choose to enable or disable. The dependency management system is designed to clearly inform users about what they're installing.

2. About language-specific features: OpenCC is primarily used here to enhance the Model Debug Panel functionality, which is crucial for identifying issues in models. The character conversion helps detect and fix encoding problems in Japanese morph names, not to specifically support Chinese.

3. About implementation strategy: I considered integrating only the necessary parts of OpenCC directly into mmd_tools (avoiding pip), but this would hinder future extensibility. My vision is to build a foundation that could later support more powerful translation capabilities through libraries like Google Translate or offline translation tools.

I believe the Model Debug Panel is an essential core feature for mmd_tools. Users need to identify issues in their downloaded/modified/created models. While I haven't added many features to it yet, it's a critical first step in solving common model problems.

I'm open to discussing alternative approaches that maintain extensibility while addressing your concerns. Perhaps we could:

1. Better document the optional nature of these dependencies
2. Create a more modular extension system where translation enhancements can be loaded through extension points
3. Start with minimal functionality that doesn't require external dependencies, then gradually introduce more advanced features

What do you think would be the best path forward?

UuuNyaa commented on mmd_tools/operators/model_validation.py:
```diff
@@ -0,0 +1,745 @@
+
+
+# Configure logging
+logging.basicConfig(level=logging.INFO, format="%(levelname)s: [%(name)s] %(message)s")
```
`basicConfig` should not be called because of its effect on the entire Python interpreter.

UuuNyaa commented:
@rintrint Thank you, for your detailed explanation.

1. I appreciate your insights into the necessity of the Model Debug Panel. It’s clear how crucial it is for identifying and resolving issues in models.
2. I now understand that OpenCC is being used for character conversion to address encoding issues, rather than specifically supporting Chinese. It's also reassuring to know that OpenCC is a well-known library.
3. Personally, I lean towards adopting the more conservative option (3) you mentioned. In the future, I hope we can explore options such as:
   - Relying on web APIs for translation
   - Utilizing Blender's potential support for LLM APIs
   - Depending on local translation libraries

I would also be interested in hearing the opinions of other contributors on this matter.

rintrint commented:
Another approach is to use opencc-python, which can be directly included in mmd_tools without needing pip.

This works only because opencc happens to have a pure Python implementation.
Doing it this way would also block the possibility of using other powerful libraries in the future that don’t have pure Python implementations.

However, I think we can first consider solutions that don’t require pip.

rintrint commented:
@UuuNyaa After considering the points raised, I've decided to remove the OpenCC support from this PR and have created two separate PRs to address this functionality:

- #213
- #214

This approach should help reduce the scope of changes in the current PR and make it easier to merge. The core functionality of the Model Debug Panel can now be reviewed independently.

Regarding external dependencies, whether to include them and if pip is the appropriate method for installation, these discussions can continue separately through the dedicated PRs. This will allow us to thoroughly evaluate the different implementation approaches without blocking the main features.

I believe this separation of concerns will lead to a smoother review process. Looking forward to your thoughts on this approach.

UuuNyaa reviewed (approved):
✅ No security issues or malicious code.
⚠️ I have not tested it.

UuuNyaa commented:
@rintrint Thank you for limiting the scope of the PR.
This is a great idea as it allows us to focus on the core functionality of the Model Debug Panel without being sidetracked by the dependency discussions.

Let's continue to explore the options regarding external dependencies in the separate PRs you created.
This approach will enable us to thoroughly evaluate each implementation strategy.
